### PR TITLE
feat: define optional and disabled by default external signature endpoints feature

### LINKED ISSUE #722: OID4VP - New endpoints for client-provided signatures
**Describe the solution you'd like**
- add as an optional feature (default set to `false`) the new endpoints
- add a new endpoint:
  - Inputs: matched credential list, wallet did (needed?) and resolved presentation request.
  - Outputs: vp token header and payload parameters
- add a new endpoint:
  - Inputs: signed and compact serialized vp token and resolved presentation request (other params needed?).
  - Outputs: same as current version of `/usePresentationRequest` endpoint
- make it stateless, i.e., do not store any information in the wallet's database

### PR DESCRIPTION
**Pull request by cpatsonakis**
## Description

Provide a clear and concise description of the changes made in this pull request.

Fixes #722 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist

- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one phrase about this update.
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+                    description = "Collection of parameters that are necessary to invoke the submit endpoint. " +
+                            "The client is expected to, in between, sign the vp token based on the " +
+                            "vpTokenParams object that is contained within."
+                    body<PrepareOID4VPResponse> {
+                        required = true
+                    }
+                }
+            }
+        }) {
+            val logger = KotlinLogging.logger { }
```
Comment (mikeplotean) [inline waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt:81]: could be moved to 'global' context, e.g.
```kotlin
fun Application.exchangeExternalSignatures() = walletRoute {
    val logger = KotlinLogging.logger { }
   //...
}
```

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+                            if (responseBody != null) "Presentation failed:\n $responseBody"
+                            else "Presentation failed",
+                            redirectUri = ""
+                        )
+                    )
+                }
+            }
+
+            if (result.isSuccess) {
+//                wallet.addOperationHistory(
```
Comment (mikeplotean) [inline waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt:316]: perhaps cleanup and have a ticket for it to implement event logging for external-signature endpoints

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/FeatureCatalog.kt [hunk 3]
```diff
@@ -66,5 +72,6 @@ object FeatureCatalog : ServiceFeatureCatalog {
         runtimeMockFeature,
         didWebRegistry,
         x5cAuthFeature,
```
Comment (mikeplotean) [inline waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/FeatureCatalog.kt:74]: just wondering if there are use cases where only one of these features is used
would it make sense to combine them into single feature that would activate both the auth and the exchange endpoints?

**Review by mikeplotean — APPROVED** [review]

File: waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+    private suspend fun issueCredentialToWallet() {
+        lateinit var offerURL: String
+        issuerApi.issue(
+            Json.decodeFromString<IssuanceRequest>(loadResource("issuance/openbadgecredential-issuance-request.json"))
+        ) {
+            offerURL = it
+            println("offer: $offerURL")
+        }
+        var response = client.post("/wallet-api/wallet/$walletId/exchange/useOfferRequest") {
```
[the remark this answers is not in the document]
Comment (cpatsonakis) [reply waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt:172]: will do in the upcoming commit.

File: waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            setBody(
+                PrepareOID4VPRequest(
+                    did = holderDID,
+                    presentationRequestURL,
+                    matchedCredentialList.map { it.id },
+                )
+            )
+        }.expectSuccess()
+        val prepareResponse = response.body<PrepareOID4VPResponse>()
+        //na kanoume sign edw
```
[the remark this answers is not in the document]
Comment (cpatsonakis) [reply waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt:238]: oops!!! will be updated accordingly in the upcoming commit :smile:

File: waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+            )
+        }.expectSuccess()
+        verifierSessionApi.get(verificationID) {
+            assert(it.tokenResponse?.vpToken?.jsonPrimitive?.contentOrNull?.expectLooksLikeJwt() != null) { "Received no valid token response!" }
+            assert(it.tokenResponse?.presentationSubmission != null) { "should have a presentation submission after submission" }
+
+            assert(it.verificationResult == true) { "overall verification should be valid" }
+            it.policyResults.let {
+                require(it != null) { "policyResults should be available after running policies" }
+                assert(it.size > 1) { "no policies have run" }
```
[the remark this answers is not in the document]
Comment (cpatsonakis) [reply waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt:263]: the verifier API by default checks two policies, holder binding and signature on the credentials included in the vp (if i am not mistaken).

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/service/oidc4vc/CredentialFilterUtils.kt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package id.walt.webwallet.service.oidc4vc
+
+import id.walt.oid4vc.data.CredentialFormat
+import id.walt.webwallet.db.models.WalletCredential
+
+object CredentialFilterUtils {
```
[the remark this answers is not in the document]
Comment (cpatsonakis) [reply waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/service/oidc4vc/CredentialFilterUtils.kt:6]: it's only limited to this particular package and will probably be refactored/removed or w/e when the SD-JWT stuff gets updated.

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt (hunk 2 above)
Comment (cpatsonakis) [reply waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt:316]: I already have this is a to do task in the respective notion track.

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/FeatureCatalog.kt (hunk 3 above)
Comment (cpatsonakis) [reply waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/FeatureCatalog.kt:74]: in principle, the x5c auth feature and the external signatures feature do not go together. However, the oid4vci endpoints will be added in this feature.

**Review by mikeplotean — APPROVED** [review]

**Review by mikeplotean — APPROVED** [review]

**Pull request merged**